homeassistant/components/tessie/coordinator.py

```diff
@@ -0,0 +1,82 @@
+"""Tessie Data Coordinator."""
+from datetime import timedelta
+from http import HTTPStatus
+import logging
+from typing import Any
+
+from aiohttp import ClientResponseError
+from tessie_api import get_state
+
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import TessieStatus
+
+# This matches the update interval Tessie performs server side
+TESSIE_SYNC_INTERVAL = 10
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class TessieDataUpdateCoordinator(DataUpdateCoordinator):
+    """Class to manage fetching data from the Tessie API."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        api_key: str,
+        vin: str,
+        data: dict[str, Any],
+    ) -> None:
+        """Initialize Tessie Data Update Coordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name="Tessie",
+            update_method=self.async_update_data,
+            update_interval=timedelta(seconds=TESSIE_SYNC_INTERVAL),
+        )
+        self.api_key = api_key
+        self.vin = vin
+        self.session = async_get_clientsession(hass)
+        self.data = self._flattern(data)
+
+    async def async_update_data(self) -> dict[str, Any]:
+        """Update vehicle data using Tessie API."""
+        try:
+            vehicle = await get_state(
+                session=self.session,
+                api_key=self.api_key,
+                vin=self.vin,
+                use_cache=False,
```

Comment:
Its unclear from the docs but the 500s might be caused by use_cache=False

https://developer.tessie.com/reference/get-state

Maybe only the first request should be False and once its up and going True?

Reply:
Seems to be stable with 
```diff
diff --git a/homeassistant/components/tessie/coordinator.py b/homeassistant/components/tessie/coordinator.py
index 769fb62e994..e05c2b08cae 100644
--- a/homeassistant/components/tessie/coordinator.py
+++ b/homeassistant/components/tessie/coordinator.py
@@ -41,6 +41,7 @@ class TessieDataUpdateCoordinator(DataUpdateCoordinator):
         self.vin = vin
         self.session = async_get_clientsession(hass)
         self.data = self._flattern(data)
+        self.did_first_update = False
 
     async def async_update_data(self) -> dict[str, Any]:
         """Update vehicle data using Tessie API."""
@@ -49,7 +50,7 @@ class TessieDataUpdateCoordinator(DataUpdateCoordinator):
                 session=self.session,
                 api_key=self.api_key,
                 vin=self.vin,
-                use_cache=False,
+                use_cache=self.did_first_update,
             )
         except ClientResponseError as e:
             if e.status == HTTPStatus.REQUEST_TIMEOUT:
@@ -59,6 +60,8 @@ class TessieDataUpdateCoordinator(DataUpdateCoordinator):
             # Reauth will go here
             raise e
 
+
+        self.did_first_update = True
         if vehicle["state"] == TessieStatus.ONLINE:
             # Vehicle is online, all data is fresh
             return self._flattern(vehicle)
```

Reply:
So my original implementation used this, but it never knew when the car was asleep or offline as the API always replied with an old cached response. If we already have cached data in home assistant I dont care if the API cant get us newer data, as we already have the cached data.

The API call made in init does use cache, so the first call always starts us off with data.

Reply:
Do you think your changes are required given its passed cached data using `self.data = self._flattern(data)`.

I have no problem adding this code in, I just dont understand why it changes anything.

Reply:
```
2023-12-08 14:45:56.771 ERROR (MainThread) [homeassistant.components.tessie.coordinator] Update failed, unable to connect to Tessie: 500, message='Internal Server Error', url=URL('https://api.tessie.com/vin1/state?use_cache=false') (Internal Server Error)
2023-12-08 14:45:56.771 ERROR (MainThread) [homeassistant.components.tessie.coordinator] Error requesting Tessie data: 500, message='Internal Server Error', url=URL('https://api.tessie.com/vin1/state?use_cache=false')
2023-12-08 14:45:56.805 ERROR (MainThread) [homeassistant.components.tessie.coordinator] Update failed, unable to connect to Tessie: 500, message='Internal Server Error', url=URL('https://api.tessie.com/vin3/state?use_cache=false') (Internal Server Error)
2023-12-08 14:45:56.805 ERROR (MainThread) [homeassistant.components.tessie.coordinator] Error requesting Tessie data: 500, message='Internal Server Error', url=URL('https://api.tessie.com/vin6/state?use_cache=false')
2023-12-08 14:45:57.080 ERROR (MainThread) [homeassistant.components.tessie.coordinator] Update failed, unable to connect to Tessie: 500, message='Internal Server Error', url=URL('https://api.tessie.com/vin4/state?use_cache=false') (Internal Server Error)
2023-12-08 14:45:57.080 ERROR (MainThread) [homeassistant.components.tessie.coordinator] Error requesting Tessie data: 500, message='Internal Server Error', url=URL('https://api.tessie.com/vin5/state?use_cache=false')
```

It seems to be very unstable with False ...

Maybe it needs to try `False` and than if it gets a 500 try `True`?

Reply:
Ive added this in.